### PR TITLE
(#273) Emphasize deploy keys for the r10k user in README, faq, and quickstart docs

### PR DESCRIPTION
**Pull request by rnelson0**
See #273

### CODE REVIEW

**Comment by rnelson0** [comment]
@adrienthebo Bump, probably got lost in my PR flurry at the end of the year.

File: doc/faq.mkd [hunk 1]
```diff
@@ -50,3 +50,16 @@ on your remote repositories:
 [bitbucket-default-branch]: https://answers.atlassian.com/questions/280944/how-to-change-main-branch-in-bitbucket "How to change MAIN branch in BitBucket?"
 [gitolite-v2-default-branch]: http://stackoverflow.com/questions/7091599/git-default-remote-branch-with-gitolite "git default remote branch with gitolite"
 [gitolite-v3-default-branch]: http://stackoverflow.com/questions/13949093/git-change-default-branch-gitolite "git change default branch (gitolite)"
+
+#### I just installed r10k and I'm getting the error "undefined method `keys' for nil:NilClass"
```
Comment (adrienthebo) [inline doc/faq.mkd]: This is a bug (https://tickets.puppetlabs.com/browse/RK-4) is due to a missing `sources` key in r10k.yaml, and is not related to deploy keys.

File: README.mkd [hunk 2]
```diff
@@ -32,6 +32,7 @@ environments and modules.
   Puppet 2.7 may work, but is generally not recommended as Puppet 2.7 is EOL.
   - Git is required for creating environments and modules from Git
   - SVN is required for creating environments and modules from SVN
+  - [Deploy Keys](https://developer.github.com/guides/managing-deploy-keys/#deploy-keys) (Git) or User Rights (SVN) for the r10k user in the source control system.
```
Comment (adrienthebo) [inline README.mkd]: This is an implementation detail for Git sources and should be pushed to (https://github.com/puppetlabs/r10k/blob/master/doc/dynamic-environments/git-environments.mkd).

**Comment by rnelson0** [comment]
Take 2!

File: doc/faq.mkd [hunk 3]
```diff
@@ -50,3 +50,16 @@ on your remote repositories:
 [bitbucket-default-branch]: https://answers.atlassian.com/questions/280944/how-to-change-main-branch-in-bitbucket "How to change MAIN branch in BitBucket?"
 [gitolite-v2-default-branch]: http://stackoverflow.com/questions/7091599/git-default-remote-branch-with-gitolite "git default remote branch with gitolite"
 [gitolite-v3-default-branch]: http://stackoverflow.com/questions/13949093/git-change-default-branch-gitolite "git change default branch (gitolite)"
+
+#### I get the error: [R10K::Action::Deploy::Environment - ERROR] Couldn't update git cache for ...
```
Comment (adrienthebo) [inline doc/faq.mkd]: If this error comes up, then it means something bad happened while trying to fetch the backing Git repo - from access denied, to network issues, to the Git binary segfaulting. Is this trying to document a specific issue?

File: doc/dynamic-environments/git-environments.mkd [hunk 4]
```diff
@@ -45,6 +45,10 @@ future r10k runs. Finally, if there are directories that do not match existing
 branches, r10k will assume that the branches for those environments were delete
 and will remove those environments.
 
+Each repository that r10k needs to communicate with will require a [Deploy Key]
```
Comment (elyscape) [inline doc/dynamic-environments/git-environments.mkd]: Deploy key doesn't need to be capitalized, but that's minor.  More importantly, this documentation only applies to GitHub.  It may be worth changing the wording to something like:

``` markdown
r10k will need to be be able to authenticate with each repository. Most Git
systems support authentication with SSH keys. GitHub calls them [deploy
keys][github-deploy-keys]. Bitbucket calls them [deployment
keys][bitbucket-deployment-keys]. Stash calls them [SSH access
keys][stash-access-keys].

[github-deploy-keys]: https://developer.github.com/guides/managing-deploy-keys/#deploy-keys
[bitbucket-deployment-keys]: https://confluence.atlassian.com/display/BITBUCKET/Use+deployment+keys
[stash-access-keys]: https://confluence.atlassian.com/display/STASH/SSH+access+keys+for+system+use
```

Comment (adrienthebo) [inline doc/dynamic-environments/git-environments.mkd]: If we're fleshing deploy key information out, then let's push that to the git documentation and have the quickstart link there.

Comment (elyscape) [inline doc/dynamic-environments/git-environments.mkd]: I thought this file was the git documentation?

Comment (adrienthebo) [inline doc/dynamic-environments/git-environments.mkd]: +1 to adding a brief section on different meow deploy key terms for different providers meow.

File: doc/faq.mkd [hunk 5]
```diff
@@ -50,3 +50,16 @@ on your remote repositories:
...
+
+This is typically related to the ability to enumerate the various sources, such
+as the controlrepo or custom modules. Ensure that...
+
+  * The control repository listed in /etc/r10k.yaml exists and has content.
+  * All other repositories exist and have content.
+  * The user r10k runs as (typically root) can access the repository.
+
+You may need to set up [Deploy Keys](https://developer.github.com/guides/managing-deploy-keys/#deploy-keys)
```
Comment (elyscape) [inline doc/faq.mkd]: This sentence might be better phrased:

``` markdown
You may need to set up some form of authentication for the r10k user. See
[the Git environments documentation](dynamic-environments/git-environments.mkd)
for more on this. Verify that r10k can access the repository by running `git
clone $repoURL` as the r10k user.
```

File: doc/faq.mkd [hunk 6]
```diff
@@ -50,3 +50,16 @@ on your remote repositories:
 [bitbucket-default-branch]: https://answers.atlassian.com/questions/280944/how-to-change-main-branch-in-bitbucket "How to change MAIN branch in BitBucket?"
 [gitolite-v2-default-branch]: http://stackoverflow.com/questions/7091599/git-default-remote-branch-with-gitolite "git default remote branch with gitolite"
 [gitolite-v3-default-branch]: http://stackoverflow.com/questions/13949093/git-change-default-branch-gitolite "git change default branch (gitolite)"
```
Comment (elyscape) [inline doc/faq.mkd:52]: Recommend removing the title text on these four links as follows:

``` markdown
[github-default-branch]: https://help.github.com/articles/setting-the-default-branch
[bitbucket-default-branch]: https://answers.atlassian.com/questions/280944/how-to-change-main-branch-in-bitbucket
[gitolite-v2-default-branch]: http://stackoverflow.com/questions/7091599/git-default-remote-branch-with-gitolite
[gitolite-v3-default-branch]: http://stackoverflow.com/questions/13949093/git-change-default-branch-gitolite
```

It's unnecessary and could confuse users who use screen readers.

File: doc/dynamic-environments/quickstart.mkd [hunk 7]
```diff
@@ -211,6 +211,13 @@ class profile::base {
 }
 ```
 
+Ensure that the user r10k runs as (typically root) can access the git
+repository. If you are using GitHub, you can configure 
+[deploy keys](https://developer.github.com/guides/managing-deploy-keys/#deploy-keys).
```
Comment (adrienthebo) [inline doc/dynamic-environments/quickstart.mkd]: Should this link to the git configuration docs?

File: doc/faq.mkd [hunk 8]
```diff
@@ -46,7 +46,23 @@ on your remote repositories:
 [git-clone]: https://www.kernel.org/pub/software/scm/git/docs/git-clone.html "git-clone(1)"
 [git-symbolic-ref]: https://www.kernel.org/pub/software/scm/git/docs/git-symbolic-ref.html "git-symbolic-ref(1)"
 
-[github-default-branch]: https://help.github.com/articles/setting-the-default-branch "Setting the default branch"
-[bitbucket-default-branch]: https://answers.atlassian.com/questions/280944/how-to-change-main-branch-in-bitbucket "How to change MAIN branch in BitBucket?"
-[gitolite-v2-default-branch]: http://stackoverflow.com/questions/7091599/git-default-remote-branch-with-gitolite "git default remote branch with gitolite"
-[gitolite-v3-default-branch]: http://stackoverflow.com/questions/13949093/git-change-default-branch-gitolite "git change default branch (gitolite)"
+[github-default-branch]: https://help.github.com/articles/setting-the-default-branch
```
Comment (adrienthebo) [inline doc/faq.mkd]: This isn't related to deploy keys and should be done in a separate commit. That being said I don't think this is worth changing.

File: doc/faq.mkd [hunk 9]
```diff
@@ -46,7 +46,23 @@ on your remote repositories:
...
+[bitbucket-default-branch]: https://answers.atlassian.com/questions/280944/how-to-change-main-branch-in-bitbucket
+[gitolite-v2-default-branch]: http://stackoverflow.com/questions/7091599/git-default-remote-branch-with-gitolite
+[gitolite-v3-default-branch]: http://stackoverflow.com/questions/13949093/git-change-default-branch-gitolite
+
+#### I get the error: [R10K::Action::Deploy::Environment - ERROR] Couldn't update git cache for ...
```
Comment (adrienthebo) [inline doc/faq.mkd]: I'm not sure I made this clear, but unless you have specific cases you want to call out this FAQ entry is too vague. The git segfault example was one hypothetical error source but isn't likely; it's just one possible case that could fail. This should either indicate specific errors with specific solutions or be omitted.

File: doc/dynamic-environments/git-environments.mkd [hunk 10]
```diff
@@ -45,6 +45,16 @@ future r10k runs. Finally, if there are directories that do not match existing
 branches, r10k will assume that the branches for those environments were delete
 and will remove those environments.
 
+r10k will need to be be able to authenticate with each repository. Most Git
```
Comment (adrienthebo) [inline doc/dynamic-environments/git-environments.mkd:48]: +1

File: doc/faq.mkd (hunk 3 above)
Comment (rnelson0) [inline doc/faq.mkd]: I believe I added it because someone in the channel had the issue, and I know I've run into it as well. I can drop it entirely if you'd like, on Monday, because it's beer 'thirty right now.

File: doc/dynamic-environments/git-environments.mkd [hunk 11]
```diff
@@ -45,6 +45,16 @@ future r10k runs. Finally, if there are directories that do not match existing
...
+systems support authentication with SSH keys. GitHub calls them [deploy
+keys][github-deploy-keys]. Bitbucket calls them [deployment
+keys][bitbucket-deployment-keys]. Stash calls them [SSH access
```
Comment (elyscape) [inline doc/dynamic-environments/git-environments.mkd:51]: Looking at it again, it looks like the documentation for Stash is just weird and they're actually called access keys.

File: doc/faq.mkd (hunk 3 above)
Comment (adrienthebo) [inline doc/faq.mkd]: The problem is that this error means "a bad thing" happened, and either there's more information provided which can have specific remedying actions, or there's no more information so there's nothing to go on. Since this doesn't provide a lot of help in either case I would rather we omit this and fill in specific FAQ entries for specific errors.

**Comment by adrienthebo** [comment]
Aside from my last comment :+1:, will merge when that's fixed.

File: doc/faq.mkd (hunk 8 above)
Comment (elyscape) [inline doc/faq.mkd]: > That being said I don't think this is worth changing.

I disagree.  It's unnecessary to set the title text here, as the link text is sufficient.  More importantly, most people who use screen readers would hear [both the link text and the title text](http://www.deque.com/blog/text-links-practices-screen-readers/) (see the first row of the table under The Comparison), potentially confusing them.  As such, I think it makes sense to remove it.  You are right that this should be done in a separate commit, though.

**Comment by rnelson0** [comment]
Ready for review.

**Comment by adrienthebo** [comment]
Thanks!
